### PR TITLE
Fix Dns.Resolve issue by using ContinueWith and AggregateException.Flatten

### LINKED ISSUE #5154: Akka.IO.Tcp: connecting to an unreachable DnsEndpoint never times out
**Version Information**
Version of Akka.NET? v1.4.21
Which Akka.NET Modules? Akka

**Describe the bug**
When attempting to connect to an unreachable `DnsEndpoint` via Akka.IO.Tcp:

```csharp
var invalidEnpdoint = new DnsEndPoint("fake", 1000);
  var pbmClient = PetabridgeCmdClient.Get(Sys);
  var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(10));
  var client = await pbmClient.StartRemoteClient(invalidEnpdoint, cts.Token); 
```

Which makes the following internal call:

```csharp
 _clientActor = actorRefFactory.ActorOf(Props.Create(() => new RemoteActorClient(_isDisposed, connected))
                    .WithDeploy(Deploy.Local),
                "remoteClient" + _counter.GetAndIncrement());
            
// kick off the TCP connection process
// the RemoteActorClient will handle all of the appropriate success / fail  signaling upstream
tcpActor.Tell(new Tcp.Connect(_remoteEndpoint, timeout: TimeSpan.FromSeconds(5)), _clientActor);
```

The `Tcp.Connect` call never fails - we simply get an endless printout of the following:

```
ERROR][7/25/2021 11:25:35 PM][Thread 0018][akka://akkacmd/system/$a/$a] One or more errors occurred.
Cause: System.AggregateException: One or more errors occurred. ---> System.Net.Sockets.SocketException: No such host is known
   at System.Net.Dns.HostResolutionEndHelper(IAsyncResult asyncResult)
   at System.Net.Dns.EndGetHostEntry(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Akka.IO.InetAddressDnsResolver.Receive(Object message)
   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   at Akka.Actor.ActorCell.ReceiveMessage(Object message)
   at Akka.Actor.ActorCell.Invoke(Envelope envelope)
---> (Inner Exception #0) System.Net.Sockets.SocketException (0x80004005): No such host is known
   at System.Net.Dns.HostResolutionEndHelper(IAsyncResult asyncResult)
   at System.Net.Dns.EndGetHostEntry(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)<---

[ERROR][7/25/2021 11:25:35 PM][Thread 0016][akka://akkacmd/system/$b/$a] One or more errors occurred.
Cause: System.AggregateException: One or more errors occurred. ---> System.Net.Sockets.SocketException: No such host is known
   at System.Net.Dns.HostResolutionEndHelper(IAsyncResult asyncResult)
   at System.Net.Dns.EndGetHostEntry(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Akka.IO.InetAddressDnsResolver.Receive(Object message)
   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   at Akka.Actor.ActorCell.ReceiveMessage(Object message)
   at Akka.Actor.ActorCell.Invoke(Envelope envelope)
---> (Inner Exception #0) System.Net.Sockets.SocketException (0x80004005): No such host is known
   at System.Net.Dns.HostResolutionEndHelper(IAsyncResult asyncResult)
   at System.Net.Dns.EndGetHostEntry(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)<---

[INFO][7/25/2021 11:25:37 PM][Thread 0039][akka://akkacmd/system/IO-TCP/$c] Message [ReceiveTimeout] from [akka://akkacmd/system/IO-TCP/$c#429187716] to [akka://akkacmd/system/IO-TCP/$c#429187716] was unhandled. [3] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO][7/25/2021 11:25:37 PM][Thread 0039][akka://akkacmd/system/IO-TCP/$b] Message [ReceiveTimeout] from [akka://akkacmd/system/IO-TCP/$b#1946559680] to [akka://akkacmd/system/IO-TCP/$b#1946559680] was unhandled. [4] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
[INFO][7/25/2021 11:25:42 PM][Thread 0037][akka://akkacmd/system/IO-TCP/$b] Message [ReceiveTimeout] from [akka://akkacmd/system/IO-TCP/$b#1946559680] to [akka://akkacmd/system/IO-TCP/$b#1946559680] was unhandled. [5] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```

**Expected behavior**
The `Tcp.Connect` operation should yield a corresponding `Tcp.CommandFailed` event after the elapsed timeout.

**Actual behavior**
We never received a `Tcp.CommandFailed` event at all - just an endless loop of deadletters from `ReceiveTimeout`.

**Environment**
.NET Core 3.1

**Additional context**

If I change the `EndPoint` to an invalid `IpEndPoint` instead of a `DnsEndpoint`, then the `Tcp.Connect` operation returns the expected `Tcp.CommandFailed` event.

### PR DESCRIPTION
**Pull request by Arkatufus**
Closes #5154
~~Fix option to #5261 (pick one)~~ This fix will be implemented.

### CODE REVIEW

File: src/core/Akka/IO/InetAddressDnsResolver.cs [hunk 1]
```diff
@@ -43,28 +45,39 @@ public InetAddressDnsResolver(SimpleDnsCache cache, Config config)
...
-                }
-                Sender.Tell(answer);
+                    
+                    newAnswer = Dns.Resolved.Create(resolve.Name, t.Result.AddressList.Where(x => 
+                        x.AddressFamily == AddressFamily.InterNetwork 
+                        || _useIpv6 && x.AddressFamily == AddressFamily.InterNetworkV6));
+                    _cache.Put(newAnswer, _positiveTtl);
+                    return newAnswer;
+
+                }, TaskContinuationOptions.ExecuteSynchronously).PipeTo(Sender);
```
Comment (Aaronontheweb) [inline src/core/Akka/IO/InetAddressDnsResolver.cs:80]: Need to close over `Sender` here

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
One small bug that needs to be addressed

File: src/core/Akka/IO/InetAddressDnsResolver.cs (hunk 1 above)
Comment (Arkatufus) [reply src/core/Akka/IO/InetAddressDnsResolver.cs]: Done

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Comment by Aaronontheweb** [comment]
Some of your own tests on this PR are failing

File: src/core/Akka/IO/InetAddressDnsResolver.cs [hunk 2]
```diff
@@ -43,28 +45,40 @@ public InetAddressDnsResolver(SimpleDnsCache cache, Config config)
...
                     {
-                        if (ex.SocketErrorCode != SocketError.HostNotFound) throw;
-                         answer = new Dns.Resolved(resolve.Name, Enumerable.Empty<IPAddress>(), Enumerable.Empty<IPAddress>());
-                        _cache.Put(answer, _negativeTtl);
+                        foreach (var exception in t.Exception.Flatten().InnerExceptions)
+                        {
+                            if (exception is SocketException se && se.SocketErrorCode == SocketError.HostNotFound)
+                            {
+                                newAnswer = new Dns.Resolved(resolve.Name, Enumerable.Empty<IPAddress>(), Enumerable.Empty<IPAddress>());
+                                _cache.Put(newAnswer, _negativeTtl);
```
Comment (Aaronontheweb) [inline src/core/Akka/IO/InetAddressDnsResolver.cs:68]: Don't cache failures, since they can be intermittent

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Need to handle all DNS exceptions and not cache failures

File: src/core/Akka/IO/Dns.cs [hunk 3]
```diff
@@ -103,24 +110,28 @@ public class Resolved : Command
...
             /// <summary>
             /// TBD
             /// </summary>
-            public IEnumerable<IPAddress> Ipv4 { get; private set; }
+            public IImmutableList<IPAddress> Ipv4 { get; }
             /// <summary>
             /// TBD
             /// </summary>
-            public IEnumerable<IPAddress> Ipv6 { get; private set; }
+            public IImmutableList<IPAddress> Ipv6 { get; }
```
Comment (Aaronontheweb) [inline src/core/Akka/IO/Dns.cs:134]: This is a public API change - may as well leave it as `IEnumerable`

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Have an API change in here that we probably can leave alone - it's minor, but the API approval tests are failing

File: src/core/Akka/IO/Dns.cs [hunk 4]
```diff
@@ -103,24 +110,28 @@ public class Resolved : Command
...
+                Ipv4 = ipv4?.ToImmutableList() ?? ImmutableList<IPAddress>.Empty;
+                Ipv6 = ipv6?.ToImmutableList() ?? ImmutableList<IPAddress>.Empty;
 
-                _addr = ipv4.FirstOrDefault() ?? ipv6.FirstOrDefault();
+                _addr = Ipv4.FirstOrDefault() ?? Ipv6.FirstOrDefault();
             }
 
+            public bool IsSuccess => Exception == null;
+            
+            public Exception Exception { get; }
```
Comment (Arkatufus) [inline src/core/Akka/IO/Dns.cs:121]: Need to add this new property to pass back exceptions from failed DNS resolve call

File: src/core/Akka/IO/Dns.cs [hunk 5]
```diff
@@ -129,8 +140,11 @@ public IPAddress Addr
             {
                 get
                 {
-                    //TODO: Throw better exception
-                    if (_addr == null) throw new Exception("Unknown host");
+                    if(Exception != null)
+                        ExceptionDispatchInfo.Capture(Exception).Throw();
```
Comment (Arkatufus) [inline src/core/Akka/IO/Dns.cs:144]: Failed DNS resolve exception is being rethrown here

**Pull request merged**